apps/elixir_ls_utils/lib/config_parser.ex

```diff
@@ -0,0 +1,171 @@
+      doc: "Mix environment to use for compilation"
+    },
+    %SettingDef{
+      key: :mix_target,
+      json_key: "mixTarget",
+      type: :string,
+      default: "host",
+      doc: "Mix target (`MIX_TARGET`) to use for compilation (requires Elixir >= 1.8)"
+    },
+    %SettingDef{
+      key: :project_dir,
+      json_key: "projectDir",
+      type: :string,
+      default: "",
+      doc:
+        "Subdirectory containing Mix project if not in the project root. " <>
+          "If value is \"\" then defaults to the workspace rootUri."
+    },
+    %SettingDef{
+      key: :suggest_specs,
+      json_key: "suggestSpecs",
+      type: :boolean,
+      default: true,
+      doc:
+        "Suggest @spec annotations inline using Dialyzer's inferred success typings " <>
+          "(Requires Dialyzer)"
+    },
+    %SettingDef{
+      key: :trace,
+      json_key: "trace",
+      type: :map,
+      default: %{},
+      doc: "Ignored"
+    }
+  ]
+
+  def load_config_file(path) do
+    with {:ok, contents} <- File.read(path) do
+      load_config(contents)
+    end
+  end
+
+  def load_config(contents) do
+    with {:ok, settings_map} <- json_decode(contents),
+         {:ok, validated_options} <- parse_config(settings_map) do
+      {:ok, Map.new(validated_options), []}
+    end
+  end
+
+  def default_config do
+    @settings
+    |> Map.new(fn %SettingDef{} = setting_def ->
+      %SettingDef{key: key, default: default} = setting_def
+      {key, default}
+    end)
+  end
+
+  @doc """
+  Parse the raw decoded JSON to the settings map (including translation from
+  camelCase to snake_case)
+  """
+  def parse_config(settings_map) do
+    # Because we use a configuration layering approach, this configuration
+    # parsing should be based on the settings_map and not the possible settings.
+    # The return value should be *only* the settings that were passed in, don't
+    # return the defaults here.
+    values =
+      settings_map
+      |> Enum.map(fn {json_key, value} ->
+        case translate_key(json_key) do
+          {:ok, key} -> {:ok, {key, value}}
+          {:error, "unknown key"} -> {:error, {:unrecognized_configuration_key, json_key, value}}
+        end
+      end)
+
+    {good, errors} = Enum.split_with(values, &match?({:ok, _}, &1))
+    config = Map.new(good, fn {:ok, {key, val}} -> {key, val} end)
+
+    {:ok, config, errors}
+  end
+
+  for %SettingDef{key: key, json_key: json_key} <- @settings do
+    defp translate_key(unquote(json_key)) do
+      {:ok, unquote(key)}
+    end
+  end
+
+  defp translate_key(_), do: {:error, "unknown key"}
+
+  for setting <- @settings do
+    def valid_key?(unquote(setting.json_key)), do: true
+  end
+
+  def valid_key?(_), do: false
+
+  def json_decode(contents) when is_binary(contents) do
+    contents
+    |> String.split(["\n", "\r", "\r\n"], trim: true)
+    |> Enum.map(&String.trim/1)
+    # Ignore json comments
+    |> Enum.reject(&String.starts_with?(&1, "#"))
```

Comment:
vscode jsonc (JSON with comments) supports JS style comments only (`//` and `/**/`), see https://code.visualstudio.com/docs/languages/json#_json-with-comments
https://github.com/Microsoft/node-jsonc-parser

Reply:
haha, good catch. I've obviously haven't been doing very much js development recently. However since we're not using a full processor I don't think it's feasible to support multi line comments.

Reply:
I agree. A simple full line comment support is good enough.